include/mneme/displs.hpp

```diff
@@ -0,0 +1,94 @@
+#ifndef MNEME_DISPLS_H
+#define MNEME_DISPLS_H
+
+#include <cassert>
+#include <cstddef>
+#include <iterator>
+#include <utility>
+#include <vector>
+
+namespace mneme {
+
+template <typename IntT> class DisplsIterator {
+public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = std::pair<std::size_t, std::size_t>;
+    using pointer = value_type*;
+    using reference = value_type&;
+
+    DisplsIterator(std::vector<IntT> const& displs, bool end = false) : displs(displs), p(0), i(0) {
+        if (end) {
+            p = displs.size() - 1;
+            i = displs[p];
+        } else {
+            i = displs[0];
+            next();
+        }
+    }
+
+    bool operator!=(DisplsIterator const& other) {
+        return i != other.i || p != other.p || &displs != &other.displs;
+    }
+    bool operator==(DisplsIterator const& other) { return !(*this != other); }
+
+    DisplsIterator& operator++() {
+        assert(i >= displs[p] && i < displs[p + 1]);
+        ++i;
+        next();
+        return *this;
+    }
+    DisplsIterator operator++(int) {
+        DisplsIterator copy(*this);
+        ++(*this);
+        return copy;
+    }
+
+    value_type operator*() { return std::make_pair(p, i); }
+
+private:
+    void next() {
+        while (i >= displs[p + 1] && p < displs.size() - 1) {
+            ++p;
+            i = displs[p];
+        }
+    }
+
+    std::vector<IntT> const& displs;
+    std::size_t p, i;
+};
+
+template <typename IntT> class Displs {
+public:
+    Displs() : displs(1, 0) {}
+
+    Displs(std::vector<IntT> const& count) { make(count); }
+
+    void make(std::vector<IntT> const& count) {
+        displs.resize(count.size() + 1);
+        displs[0] = 0;
+        for (IntT p = 0; p < count.size(); ++p) {
```

Comment:
This is going to give a warning if IntT is signed - so we better cast count.size to the proper type?